src/rust/engine/src/watch.rs

```diff
@@ -0,0 +1,198 @@
+          format!(
+            "Failed to begin recursively watching files in the build root: {}",
+            e
+          )
+        })?
+    }
+    let wrapped_watcher = Arc::new(Mutex::new(watcher));
+
+    let (thread_liveness_sender, thread_liveness_receiver) = crossbeam_channel::unbounded();
+    thread::spawn(move || {
+      logging::set_thread_destination(logging::Destination::Pantsd);
+      loop {
+        let event_res = watch_receiver.recv_timeout(Duration::from_millis(100));
+        let graph = if let Some(g) = graph.upgrade() {
+          g
+        } else {
+          // The Graph has been dropped: we're done.
+          break;
+        };
+        match event_res {
+          Ok(Ok(ev)) => {
+            let paths: HashSet<_> = ev
+              .paths
+              .into_iter()
+              .map(|path| {
+                // relativize paths to build root.
+                let mut paths_to_invalidate: Vec<PathBuf> = vec![];
+                let path_relative_to_build_root = {
+                  if path.starts_with(&canonical_build_root) {
+                    path.strip_prefix(&canonical_build_root).unwrap().into()
+                  } else {
+                    path
+                  }
+                };
+                paths_to_invalidate.push(path_relative_to_build_root.clone());
+                if let Some(parent_dir) = path_relative_to_build_root.parent() {
+                  paths_to_invalidate.push(parent_dir.to_path_buf());
+                }
+                paths_to_invalidate
+              })
+              .flatten()
+              .collect();
+            debug!("notify invalidating {:?} because of {:?}", paths, ev.kind);
+            InvalidationWatcher::invalidate(&graph, &paths, "notify");
+          }
+          Ok(Err(err)) => {
+            if let notify::ErrorKind::PathNotFound = err.kind {
+              warn!("Path(s) did not exist: {:?}", err.paths);
+              continue;
+            } else {
+              error!("File watcher failing with: {}", err);
+              break;
+            }
+          }
+          Err(RecvTimeoutError::Timeout) => continue,
+          Err(RecvTimeoutError::Disconnected) => {
+            // The Watcher is gone: we're done.
+            break;
+          }
+        };
+      }
+      debug!("Watch thread exiting.");
+      // Signal that we're exiting (which we would also do by just dropping the channel).
+      let _ = thread_liveness_sender.send(());
+    });
+
+    Ok(InvalidationWatcher {
+      watcher: wrapped_watcher,
+      executor,
+      liveness: thread_liveness_receiver,
+      current_platform,
+    })
+  }
+
+  ///
+  /// Watch the given path non-recursively.
+  ///
+  pub async fn watch(&self, path: PathBuf) -> Result<(), notify::Error> {
+    // Short circuit here if we are on a Darwin platform because we should be watching
+    // the entire build root recursively already.
+    if self.current_platform == Platform::Darwin {
```

Comment:
These can be done using compile time conditionals, but not a blocker.

Reply:
will fix in the follow up.